File: Data Structures & Algorithms/binary-tree-diameter/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.result = 0
        def dfs(root) -> int:
            if root == None:
                return 0
            elif root.left == None and root.right == None:
                return 1
            
            leftpath = dfs(root.left)
            rightpath = dfs(root.right)
            
            self.result = max(self.result, leftpath + rightpath)

            if leftpath > rightpath:
                return leftpath + 1
            return rightpath + 1
    
        dfs(root)
        return self.result

File: Data Structures & Algorithms/binary-tree-inorder-traversal/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        if (root == null){
            return new ArrayList<Integer>();
        }
        List<Integer> result = new ArrayList<>();
        if (root.left == null && root.right == null){
            result.add(root.val);
            return result;
        } 
        
        if (root.left != null){
            List<Integer> leftTree = inorderTraversal(root.left);
            result.addAll(leftTree);
        }

        result.add(root.val);

        if (root.right != null){
            List<Integer> rightTree = inorderTraversal(root.right);
            result.addAll(rightTree);
        }

        return result;

    }
}